# Capstone 3: Spotify Customer Churn Analysis

## Business Problem

Customer churn is a significant challenge for subscription-based businesses such as Spotify. Understanding which customer characteristics are associated with churn can help the company improve retention strategies and reduce revenue loss.

The objective of this project is to prepare and clean a Spotify customer dataset for future exploratory data analysis and predictive modeling.

This notebook focuses specifically on Data Wrangling, including:

- Data acquisition
- Data organization
- Data definition
- Data cleaning
- Handling missing values
- Handling duplicates
- Identifying potential outliers

# Data Acquisition

The dataset was obtained from a public Spotify customer churn dataset.

The data is loaded directly into Python using pandas to minimize manual intervention and ensure reproducibility.

In [1]:
# Import required libraries

import pandas as pd
import numpy as np

# Load dataset

spotify_df = pd.read_csv("spotify_churn_dataset.csv")

# Preview first rows

spotify_df.head()

,user_id,gender,age,country,subscription_type,listening_time,songs_played_per_day,skip_rate,device_type,ads_listened_per_week,offline_listening,is_churned
0,1,Female,54,CA,Free,26,23,0.20,Desktop,31,0,1
1,2,Other,33,DE,Family,141,62,0.34,Web,0,1,0
2,3,Male,38,AU,Premium,199,38,0.04,Mobile,0,1,1
3,4,Female,22,CA,Student,36,2,0.31,Mobile,0,1,0
4,5,Other,29,US,Family,250,57,0.36,Mobile,0,1,1


In [2]:
spotify_df.shape

(8000, 12)

In [3]:
spotify_df.columns

Index(['user_id', 'gender', 'age', 'country', 'subscription_type',
       'listening_time', 'songs_played_per_day', 'skip_rate', 'device_type',
       'ads_listened_per_week', 'offline_listening', 'is_churned'],
      dtype='object')

In [4]:
spotify_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   user_id                8000 non-null   int64  
 1   gender                 8000 non-null   object 
 2   age                    8000 non-null   int64  
 3   country                8000 non-null   object 
 4   subscription_type      8000 non-null   object 
 5   listening_time         8000 non-null   int64  
 6   songs_played_per_day   8000 non-null   int64  
 7   skip_rate              8000 non-null   float64
 8   device_type            8000 non-null   object 
 9   ads_listened_per_week  8000 non-null   int64  
 10  offline_listening      8000 non-null   int64  
 11  is_churned             8000 non-null   int64  
dtypes: float64(1), int64(7), object(4)
memory usage: 750.1+ KB


# Data Organization

The dataset contains 8,000 Spotify users and 12 variables describing customer demographics, listening behavior, subscription characteristics, and churn status.

The dataset contains:

- Numerical variables
- Categorical variables
- Binary variables
- One target variable

The target variable for future analysis is:

is_churned

where:

- 0 = Customer retained
- 1 = Customer churned

The remaining variables will be treated as predictor variables.

In [5]:
# Separate columns by data type

numerical_columns = spotify_df.select_dtypes(include=['int64', 'float64']).columns

categorical_columns = spotify_df.select_dtypes(include=['object']).columns

print("Numerical Columns:")
print(numerical_columns)

print("\nCategorical Columns:")
print(categorical_columns)

Numerical Columns:
Index(['user_id', 'age', 'listening_time', 'songs_played_per_day', 'skip_rate',
       'ads_listened_per_week', 'offline_listening', 'is_churned'],
      dtype='object')

Categorical Columns:
Index(['gender', 'country', 'subscription_type', 'device_type'], dtype='object')


# Data Definition

This section verifies that each variable contains reasonable values and has the expected data type.

The following checks will be performed:

1. Verify data types
2. Review summary statistics
3. Review categorical values
4. Identify potential data quality issues

In [6]:
# Summary statistics for numerical variables

spotify_df.describe()

,user_id,age,listening_time,songs_played_per_day,skip_rate,ads_listened_per_week,offline_listening,is_churned
count,8000.00000,8000.000000,8000.000000,8000.000000,8000.000000,8000.000000,8000.000000,8000.000000
mean,4000.50000,37.662125,154.068250,50.127250,0.300127,6.943875,0.747750,0.258875
std,2309.54541,12.740359,84.015596,28.449762,0.173594,13.617953,0.434331,0.438044
min,1.00000,16.000000,10.000000,1.000000,0.000000,0.000000,0.000000,0.000000
25%,2000.75000,26.000000,81.000000,25.000000,0.150000,0.000000,0.000000,0.000000
50%,4000.50000,38.000000,154.000000,50.000000,0.300000,0.000000,1.000000,0.000000
75%,6000.25000,49.000000,227.000000,75.000000,0.450000,5.000000,1.000000,1.000000
max,8000.00000,59.000000,299.000000,99.000000,0.600000,49.000000,1.000000,1.000000


In [7]:
# Summary statistics for categorical variables

spotify_df.describe(include='object')

,gender,country,subscription_type,device_type
count,8000,8000,8000,8000
unique,3,8,4,3
top,Male,AU,Premium,Desktop
freq,2691,1034,2115,2778


In [8]:
# Review unique values for categorical variables

for column in categorical_columns:
    print(f"\n{column}")
    print(spotify_df[column].value_counts())


gender
gender
Male      2691
Female    2659
Other     2650
Name: count, dtype: int64

country
country
AU    1034
US    1032
DE    1015
IN    1011
PK     999
FR     989
UK     966
CA     954
Name: count, dtype: int64

subscription_type
subscription_type
Premium    2115
Free       2018
Student    1959
Family     1908
Name: count, dtype: int64

device_type
device_type
Desktop    2778
Web        2623
Mobile     2599
Name: count, dtype: int64


# Missing Values Analysis

Missing values can negatively impact statistical analysis and machine learning models.

The first step is to identify whether any missing values exist within the dataset.

If missing values are identified, an appropriate treatment strategy will be selected based on the variable type and business context.

In [9]:
# Count missing values by column

spotify_df.isnull().sum()

user_id                  0
gender                   0
age                      0
country                  0
subscription_type        0
listening_time           0
songs_played_per_day     0
skip_rate                0
device_type              0
ads_listened_per_week    0
offline_listening        0
is_churned               0
dtype: int64

In [10]:
# Total missing values in dataset

spotify_df.isnull().sum().sum()

np.int64(0)

## Findings

No missing values were identified within the dataset.

Because all variables contain complete information, no imputation or record removal was required.

# Duplicate Record Analysis

Duplicate records can distort statistical results and machine learning models by overrepresenting observations.

The dataset will be evaluated for duplicate rows.

In [11]:
# Check for duplicate rows

spotify_df.duplicated().sum()

np.int64(0)

In [12]:
# Display duplicates if any exist

spotify_df[spotify_df.duplicated()]

,user_id,gender,age,country,subscription_type,listening_time,songs_played_per_day,skip_rate,device_type,ads_listened_per_week,offline_listening,is_churned


## Findings

No duplicate records were identified within the dataset.

No records were removed.

# Data Quality Validation

The dataset was reviewed to ensure values fall within expected business ranges.

The following rules were evaluated:

- Age should be between 16 and 100
- Skip rate should fall between 0 and 1
- Listening time should be non-negative
- Songs played per day should be non-negative
- Ads listened per week should be non-negative
- Binary variables should contain only 0 and 1

In [13]:
# Verify age range

spotify_df['age'].min(), spotify_df['age'].max()

(16, 59)

In [14]:
# Verify skip rate range

spotify_df['skip_rate'].min(), spotify_df['skip_rate'].max()

(0.0, 0.6)

In [15]:
# Verify binary fields

print(spotify_df['offline_listening'].unique())

print(spotify_df['is_churned'].unique())

[0 1]
[1 0]


# Outlier Detection

Outliers were evaluated using the Interquartile Range (IQR) method.

Outliers were not automatically removed because extreme customer behavior may represent legitimate business activity.

The objective is to identify unusual observations and determine whether they are data errors or valid customer behaviors.

In [16]:
# Numerical columns to evaluate

numeric_cols = [
    'age',
    'listening_time',
    'songs_played_per_day',
    'skip_rate',
    'ads_listened_per_week'
]

for col in numeric_cols:

    Q1 = spotify_df[col].quantile(0.25)
    Q3 = spotify_df[col].quantile(0.75)

    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = spotify_df[
        (spotify_df[col] < lower_bound) |
        (spotify_df[col] > upper_bound)
    ]

    print(f"{col}: {len(outliers)} outliers")

age: 0 outliers
listening_time: 0 outliers
songs_played_per_day: 0 outliers
skip_rate: 0 outliers
ads_listened_per_week: 1683 outliers


## Outlier Findings

Potential outliers were evaluated using the Interquartile Range (IQR) method.

No outliers were identified for:

- age
- listening_time
- songs_played_per_day
- skip_rate

The variable ads_listened_per_week contained 1,683 observations identified as potential outliers.

These values were not removed because they appear to represent legitimate customer behavior rather than data entry errors. The distribution is likely influenced by differences between subscription types, where premium subscribers may receive no advertisements while free-tier users may receive substantially more.

Therefore, all observations were retained for future analysis.

# Data Cleaning Decisions

The following data quality checks were performed:

| Data Quality Issue | Findings | Action Taken |
|-------------------|-----------|-------------|
| Missing Values | No missing values identified | No action required |
| Duplicate Records | No duplicate records identified | No action required |
| Invalid Values | No invalid values identified | No action required |
| Outliers | Potential outliers identified in ads_listened_per_week | Retained as valid customer behavior |
| Data Types | All variables stored in appropriate formats | No action required |

Based on the results of the data wrangling process, the dataset was determined to be suitable for exploratory data analysis and predictive modeling.

# Export Clean Dataset

After completing the data wrangling process, the validated dataset was exported for use in subsequent exploratory data analysis and preprocessing steps.

In [17]:
# Export cleaned dataset

spotify_df.to_csv("spotify_churn_cleaned.csv", index=False)

print("Cleaned dataset exported successfully.")

Cleaned dataset exported successfully.
